ci-operator/step-registry/single-node/recert/single-node-recert-commands.sh

```diff
@@ -0,0 +1,141 @@
+      --static-dir /machine-config-daemon \
+      --use-cert /certs/admin-kubeconfig-client-ca.crt \
+      --use-key "kube-apiserver-localhost-signer /keys/localhost-serving-signer.key" \
+      --use-key "kube-apiserver-lb-signer /keys/loadbalancer-serving-signer.key" \
+      --use-key "kube-apiserver-service-network-signer /keys/service-network-serving-signer.key" \
+
+  podman kill recert_etcd
+
+  # workaround until https://github.com/omertuc/recert/blob/4d41d451ba57fbc9fd75781684906360696f384a/README.md?plain=1#L24 is resolved
+  rm -rf "/etc/machine-config-daemon/currentconfig"
+  touch "/run/machine-config-daemon-force"
+}
+
+function fetch_crts_keys {
+  mkdir -p /tmp/certs /tmp/keys
+
+  oc get cm -n openshift-config admin-kubeconfig-client-ca -ojsonpath='{.data.ca-bundle\.crt}' > /tmp/certs/admin-kubeconfig-client-ca.crt
+
+  declare -a secrets=(
+    "loadbalancer-serving-signer"
+    "localhost-serving-signer"
+    "service-network-serving-signer"
+  )
+  for secret in "${secrets[@]}"; do
+    oc get secrets -n openshift-kube-apiserver-operator "${secret}" -ojsonpath='{.data.tls\.key}' | base64 -d > "/tmp/keys/${secret}.key"
+  done
+}
+
+function delete_crts_keys {
+  rm -rf /tmp/certs /tmp/keys
+}
+
+wait_for_api
+fetch_crts_keys
+stop_containers
+
+recert
+
+start_containers
+delete_crts_keys
+
+touch /var/recert.done
+echo "Recert successfully run."
+EOF
+)
+
+# Base64 encode the script for use in the MachineConfig.
+b64_script=$(echo "${recert_script}" | base64 -w 0)
+
+machineconfig=$(oc create -f - -o jsonpath='{.metadata.name}' << EOF
+apiVersion: machineconfiguration.openshift.io/v1
+kind: MachineConfig
+metadata:
+  labels:
+    machineconfiguration.openshift.io/role: master
+  name: 99-master-recert
+spec:
+  config:
+    ignition:
+      version: 3.2.0
+    storage:
+      files:
+      - contents:
+          source: data:text/plain;charset=utf-8;base64,${b64_script}
+        mode: 493
+        overwrite: true
+        path: /usr/local/bin/recert.sh
+    systemd:
+      units:
+      - contents: |
+          [Unit]
+          Description=Regenerate certificates script
+          After=kubelet.service
+          ConditionPathExists=!/var/recert.done
+          [Service]
+          Type=oneshot
+          RemainAfterExit=yes
+          ExecStart=/usr/local/bin/recert.sh
+          [Install]
+          WantedBy=multi-user.target
+        enabled: true
+        name: recert.service
+EOF
+)
+echo "Created \"${machineconfig}\" MachineConfig"
+
+echo "Waiting for master MachineConfigPool to have condition=updating..."
+oc wait --for=condition=updating machineconfigpools master --timeout 2m
+
+echo "Waiting for master MachineConfigPool to have condition=updated..."
+until oc wait --for=condition=updated machineconfigpools master --timeout=5m &> /dev/null
+do
+  echo "Waiting for master MachineConfigPool to have condition=updated..."
+  sleep 5s
+done
+
+oc adm wait-for-stable-cluster --minimum-stable-period=5m --timeout=30m
```

Comment:
Should we have this after reboot as well?

Reply:
As far as I've tested, there are 2 expected reboots and an additional one (which I can't yet justify). I believe all reboots happen before the `master` `MachineConfigPool` has reached the `updated` state. 

Do you think the cluster being `stable` for `5m` is not good enough to ensure that no reboot or any other rollouts happen?

For completeness, here are the logs from the respective CI pod:

```
Created "99-master-recert" MachineConfig
Waiting for master MachineConfigPool to have condition=updating...
machineconfigpool.machineconfiguration.openshift.io/master condition met
Waiting for master MachineConfigPool to have condition=updated...
Waiting for master MachineConfigPool to have condition=updated...
Waiting for master MachineConfigPool to have condition=updated...
Waiting for master MachineConfigPool to have condition=updated...
Waiting for master MachineConfigPool to have condition=updated...
Waiting for master MachineConfigPool to have condition=updated...
Waiting for master MachineConfigPool to have condition=updated...
clusteroperators/authentication is unavailable and progressing and degraded at 2023-08-30T11:15:43Z
clusteroperators/kube-apiserver is progressing at 2023-08-30T11:15:43Z

clusteroperators/authentication is still unavailable and progressing and degraded after 10s
clusteroperators/kube-apiserver is still progressing after 10s
clusteroperators/network became progressing at 2023-08-30T11:15:53Z

clusteroperators/authentication is still unavailable and progressing after 20s
clusteroperators/kube-apiserver is still progressing after 20s
clusteroperators/network stabilized at 2023-08-30T11:16:03Z after 10s

failed to list clusteroperators: Get "https://api.ci-op-80d6pc6c-0bd4c.aws-2.ci.openshift.org:6443/apis/config.openshift.io/v1/clusteroperators": dial tcp 44.225.66.237:6443: connect: connection refused
failed to list clusteroperators: Get "https://api.ci-op-80d6pc6c-0bd4c.aws-2.ci.openshift.org:6443/apis/config.openshift.io/v1/clusteroperators": dial tcp 44.225.66.237:6443: connect: connection refused
failed to list clusteroperators: Get "https://api.ci-op-80d6pc6c-0bd4c.aws-2.ci.openshift.org:6443/apis/config.openshift.io/v1/clusteroperators": dial tcp 44.226.138.107:6443: connect: connection refused
failed to list clusteroperators: Get "https://api.ci-op-80d6pc6c-0bd4c.aws-2.ci.openshift.org:6443/apis/config.openshift.io/v1/clusteroperators": dial tcp 44.226.138.107:6443: i/o timeout
clusteroperators/authentication is still unavailable and progressing after 2m1s
clusteroperators/kube-apiserver is still progressing after 2m1s

clusteroperators/authentication is still unavailable and progressing after 2m10s
clusteroperators/kube-apiserver is still progressing after 2m10s

clusteroperators/authentication is still unavailable and progressing after 2m25s
clusteroperators/kube-apiserver is still progressing after 2m25s

clusteroperators/authentication is still unavailable and progressing after 2m33s
clusteroperators/kube-apiserver is still progressing after 2m33s

clusteroperators/authentication is still unavailable and progressing after 2m41s
clusteroperators/kube-apiserver is still progressing after 2m41s

clusteroperators/authentication is still unavailable and progressing after 2m51s
clusteroperators/kube-apiserver is still progressing after 2m51s

clusteroperators/authentication is still unavailable and progressing after 3m0s
clusteroperators/dns became unavailable and progressing and degraded at 2023-08-30T11:18:43Z
clusteroperators/kube-apiserver is still progressing after 3m0s
clusteroperators/operator-lifecycle-manager-packageserver became unavailable and progressing at 2023-08-30T11:18:43Z

clusteroperators/authentication is still unavailable and progressing after 3m10s
clusteroperators/dns stabilized at 2023-08-30T11:18:53Z after 10s
clusteroperators/ingress became unavailable and progressing at 2023-08-30T11:18:53Z
clusteroperators/kube-apiserver is still progressing after 3m10s
clusteroperators/operator-lifecycle-manager-packageserver stabilized at 2023-08-30T11:18:53Z after 10s

clusteroperators/authentication is still unavailable and progressing after 3m20s
clusteroperators/ingress is still unavailable and progressing after 10s
clusteroperators/kube-apiserver is still progressing after 3m20s

clusteroperators/authentication is still unavailable and progressing after 3m30s
clusteroperators/ingress is still unavailable and progressing after 20s
clusteroperators/kube-apiserver is still progressing after 3m30s

clusteroperators/authentication is still unavailable and progressing after 3m40s
clusteroperators/ingress stabilized at 2023-08-30T11:19:23Z after 30s
clusteroperators/kube-apiserver is still progressing after 3m40s

clusteroperators/authentication is still unavailable and progressing after 3m50s
clusteroperators/kube-apiserver is still progressing after 3m50s

clusteroperators/authentication is still unavailable and progressing after 4m0s
clusteroperators/kube-apiserver is still progressing after 4m0s

clusteroperators/authentication is still unavailable and progressing after 4m10s
clusteroperators/kube-apiserver is still progressing after 4m10s

clusteroperators/authentication is still unavailable and progressing after 4m20s
clusteroperators/kube-apiserver is still progressing after 4m20s

clusteroperators/authentication is still unavailable and progressing after 4m30s
clusteroperators/kube-apiserver is still progressing after 4m30s

clusteroperators/authentication is still unavailable and progressing after 4m40s
clusteroperators/kube-apiserver is still progressing after 4m40s

clusteroperators/authentication is still unavailable and progressing after 4m50s
clusteroperators/kube-apiserver is still progressing after 4m50s

clusteroperators/authentication is still unavailable and progressing after 5m0s
clusteroperators/kube-apiserver is still progressing after 5m0s

clusteroperators/authentication is still unavailable and progressing after 5m10s
clusteroperators/kube-apiserver is still progressing after 5m10s

clusteroperators/authentication is still unavailable and progressing after 5m20s
clusteroperators/kube-apiserver is still progressing after 5m20s

clusteroperators/authentication is still unavailable and progressing after 5m30s
clusteroperators/kube-apiserver is still progressing after 5m30s

clusteroperators/authentication is still unavailable and progressing after 5m40s
clusteroperators/kube-apiserver is still progressing after 5m40s

clusteroperators/authentication is still unavailable and progressing after 5m50s
clusteroperators/kube-apiserver is still progressing after 5m50s

clusteroperators/authentication is still unavailable and progressing after 6m0s
clusteroperators/kube-apiserver is still progressing after 6m0s

clusteroperators/authentication is still unavailable and progressing after 6m10s
clusteroperators/kube-apiserver is still progressing after 6m10s

clusteroperators/authentication is still unavailable and progressing after 6m20s
clusteroperators/kube-apiserver is still progressing after 6m20s

clusteroperators/authentication is still unavailable and progressing after 6m30s
clusteroperators/kube-apiserver is still progressing after 6m30s

clusteroperators/authentication is still unavailable and progressing after 6m40s
clusteroperators/kube-apiserver is still progressing after 6m40s

clusteroperators/authentication is still unavailable and progressing after 6m50s
clusteroperators/kube-apiserver is still progressing after 6m50s

clusteroperators/authentication is still unavailable and progressing after 7m0s
clusteroperators/kube-apiserver is still progressing after 7m0s

clusteroperators/authentication is still unavailable and progressing after 7m10s
clusteroperators/kube-apiserver is still progressing after 7m10s

clusteroperators/authentication is still unavailable and progressing after 7m20s
clusteroperators/kube-apiserver is still progressing after 7m20s

clusteroperators/authentication is still unavailable and progressing and degraded after 7m30s
clusteroperators/kube-apiserver is still progressing after 7m30s

clusteroperators/authentication stabilized at 2023-08-30T11:23:23Z after 7m40s
clusteroperators/kube-apiserver is still progressing after 7m40s
clusteroperators/openshift-apiserver became unavailable at 2023-08-30T11:23:23Z

clusteroperators/kube-apiserver is still progressing after 7m50s
clusteroperators/openshift-apiserver is still unavailable after 10s
clusteroperators/openshift-controller-manager became unavailable and progressing at 2023-08-30T11:23:33Z

clusteroperators/kube-apiserver is still progressing after 8m0s
clusteroperators/kube-controller-manager became progressing at 2023-08-30T11:23:43Z
clusteroperators/openshift-apiserver is still unavailable after 20s
clusteroperators/openshift-controller-manager is still unavailable and progressing after 10s

clusteroperators/kube-apiserver is still progressing after 8m10s
clusteroperators/kube-controller-manager is still progressing after 10s
clusteroperators/openshift-apiserver is still unavailable after 30s
clusteroperators/openshift-controller-manager stabilized at 2023-08-30T11:23:53Z after 20s

clusteroperators/kube-apiserver is still progressing after 8m20s
clusteroperators/kube-controller-manager is still progressing after 20s
clusteroperators/openshift-apiserver is still unavailable after 40s

clusteroperators/kube-apiserver is still progressing after 8m30s
clusteroperators/kube-controller-manager is still progressing after 30s
clusteroperators/openshift-apiserver is still unavailable after 50s

clusteroperators/kube-apiserver is still progressing after 8m40s
clusteroperators/kube-controller-manager is still progressing after 40s
clusteroperators/openshift-apiserver is still unavailable after 1m0s

clusteroperators/kube-apiserver is still progressing after 8m50s
clusteroperators/kube-controller-manager is still progressing after 50s
clusteroperators/openshift-apiserver is still unavailable after 1m10s

clusteroperators/kube-apiserver is still progressing after 9m0s
clusteroperators/kube-controller-manager is still progressing after 1m0s
clusteroperators/openshift-apiserver is still unavailable after 1m20s

clusteroperators/image-registry became progressing at 2023-08-30T11:24:53Z
clusteroperators/kube-apiserver is still progressing after 9m10s
clusteroperators/kube-controller-manager is still progressing after 1m10s
clusteroperators/kube-scheduler became progressing at 2023-08-30T11:24:53Z
clusteroperators/openshift-apiserver stabilized at 2023-08-30T11:24:53Z after 1m30s

clusteroperators/image-registry is still progressing after 10s
clusteroperators/kube-apiserver is still progressing after 9m20s
clusteroperators/kube-controller-manager is still progressing after 1m20s
clusteroperators/kube-scheduler is still progressing after 10s

clusteroperators/console became progressing at 2023-08-30T11:25:13Z
clusteroperators/image-registry stabilized at 2023-08-30T11:25:13Z after 20s
clusteroperators/kube-apiserver is still progressing after 9m30s
clusteroperators/kube-controller-manager is still progressing after 1m30s
clusteroperators/kube-scheduler is still progressing after 20s

clusteroperators/console stabilized at 2023-08-30T11:25:23Z after 10s
clusteroperators/kube-apiserver is still progressing after 9m40s
clusteroperators/kube-controller-manager is still progressing after 1m40s
clusteroperators/kube-scheduler is still progressing after 30s

clusteroperators/kube-apiserver is still progressing after 9m50s
clusteroperators/kube-controller-manager is still progressing after 1m50s
clusteroperators/kube-scheduler is still progressing after 40s

clusteroperators/kube-apiserver is still progressing after 10m0s
clusteroperators/kube-controller-manager is still progressing after 2m0s
clusteroperators/kube-scheduler is still progressing after 50s

clusteroperators/kube-apiserver is still progressing after 10m10s
clusteroperators/kube-controller-manager is still progressing after 2m10s
clusteroperators/kube-scheduler is still progressing after 1m0s

failed to list clusteroperators: Get "https://api.ci-op-80d6pc6c-0bd4c.aws-2.ci.openshift.org:6443/apis/config.openshift.io/v1/clusteroperators": dial tcp 44.226.138.107:6443: connect: connection refused
failed to list clusteroperators: Get "https://api.ci-op-80d6pc6c-0bd4c.aws-2.ci.openshift.org:6443/apis/config.openshift.io/v1/clusteroperators": dial tcp 44.226.138.107:6443: connect: connection refused
clusteroperators/kube-apiserver is still progressing after 10m47s
clusteroperators/kube-controller-manager stabilized at 2023-08-30T11:26:29Z after 2m46s
clusteroperators/kube-scheduler is still progressing after 1m36s

clusteroperators/kube-apiserver is still progressing after 10m50s
clusteroperators/kube-scheduler is still progressing after 1m40s

clusteroperators/kube-apiserver is still progressing after 11m0s
clusteroperators/kube-scheduler stabilized at 2023-08-30T11:26:43Z after 1m50s

clusteroperators/kube-apiserver is still progressing after 11m10s

clusteroperators/kube-apiserver is still progressing after 11m20s

clusteroperators/kube-apiserver stabilized at 2023-08-30T11:27:13Z after 11m30s
All clusteroperators became stable at 2023-08-30T11:27:13Z

All clusteroperators are still stable after 10s

All clusteroperators are still stable after 20s

All clusteroperators are still stable after 30s

All clusteroperators are still stable after 40s

All clusteroperators are still stable after 50s

All clusteroperators are still stable after 1m0s

All clusteroperators are still stable after 1m10s

All clusteroperators are still stable after 1m20s

All clusteroperators are still stable after 1m30s

All clusteroperators are still stable after 1m40s

All clusteroperators are still stable after 1m50s

All clusteroperators are still stable after 2m0s

All clusteroperators are still stable after 2m10s

All clusteroperators are still stable after 2m20s

All clusteroperators are still stable after 2m30s

All clusteroperators are still stable after 2m40s

All clusteroperators are still stable after 2m50s

All clusteroperators are still stable after 3m0s

All clusteroperators are still stable after 3m10s

All clusteroperators are still stable after 3m20s

All clusteroperators are still stable after 3m30s

All clusteroperators are still stable after 3m40s

All clusteroperators are still stable after 3m50s

All clusteroperators are still stable after 4m0s

All clusteroperators are still stable after 4m10s

All clusteroperators are still stable after 4m20s

All clusteroperators are still stable after 4m30s

All clusteroperators are still stable after 4m40s

All clusteroperators are still stable after 4m50s

All clusteroperators are still stable after 5m0s

All clusteroperators are still stable after 5m10s

All clusteroperators are stable
```